lib/indexer/comparer.rb

```diff
@@ -0,0 +1,122 @@
+# require_relative 'app';
+# require 'pry'
+# Indexer::Comparer.new(
+#   'government',
+#   'government-old-2017-06-21t11:04:56z-00000000-0000-0000-0000-000000000000',
+#   'government-2017-06-21t11:55:51z-00000000-0000-0000-0000-000000000000'
+# ).run
+
+module Indexer
+  class Comparer
+    # Source fields to ignore when comparing old and new objects for equality.
+    # These are ones which we know have changed for most documents.
+    FIELDS_TO_IGNORE = ["_id", "_type", "popularity"]
+
+    def initialize(base_name, old_index_name, new_index_name, rejected_fields = nil)
+      @base_name = base_name
+      @old_index_name = old_index_name
+      @new_index_name = new_index_name
+      @rejected_fields = rejected_fields || ['popularity']
+    end
+
+    def log(msg)
+      puts "#{Time.now}: #{msg}"
+    end
+
+    def read_index(index_name, filter = nil)
+      log "Access index: #{index_name}"
+      index = search_config.search_server.index_group(@base_name).index_for_name(index_name)
+
+      documents = {}
+
+
+      search_body = filter || { query: { match_all: {} } }
+      batch_size = index.class.scroll_batch_size
+
+      log "Starting to load index: #{index_name} - #{search_body.inspect}"
+
+      ScrollEnumerator.new(
+        client: index.send(:build_client, timeout_options),
+        index_names: index_name,
+        search_body: search_body,
+        batch_size: batch_size
+      ) do |document|
+        result = document['_source']
+        root_elements = document.each_with_object({}) do |(k, v), h|
+          h["_root#{k}"] = v unless k == '_source'
+        end
+        result.merge(root_elements)
+      end.each do |d|
+        key = [d['_root_id'], d['_root_type']]
+        documents[key] = d
+      end
+
+      log "Finished loading index: #{index_name}"
+
+      documents
+    end
+
+    def timeout_options
+      {
+        timeout: 30.0
+      }
+    end
+
+    def reject_fields(hash)
+      hash.reject{ |k, _| @rejected_fields.include?(k) }
+    end
+
+    def run
+      outcomes = Hash.new { |h, k| h[k] = 0 }
+
+      filters.each do |filter|
+        old_data = read_index(@old_index_name, filter)
+        new_data = read_index(@new_index_name, filter)
```

Comment:
I think the only part of `new_data` you need to keep in memory is the keys (for the `removed_keys` calculation below). So if memory usage is an issue you might be able to do a second pass with the ScrollEnumerator instead.